### PR TITLE
Handle relative paths correctly

### PR DESCRIPTION
**Pull request by raykyri**
- Fix error that occurs with relative paths
- Fix error that occurs when refreshing the page title before a chart has loaded, so that error messages show up correctly
- Remove extraneous next() statements to avoid unnecessary warnings
- Bump version num to 0.2.2.

Reviewers: @mikesall, @valueof if you'd like to take a quick scan

### CODE REVIEW

File: lib/charted.js [hunk 1]
```diff
@@ -9,16 +9,19 @@ module.exports = function(app, charted_root) {
   if (charted_root === undefined) { charted_root = '/charted/' }
 
   app.use(charted_root + 'get', function (req, res, next) {
-    var get_url = url.parse(req.url, true).query.url
```
Comment (valueof) [inline lib/charted.js]: What was the point for removing this variable if you ended up using `urlObj.query.url` everywhere else?

File: lib/charted.js [hunk 2]
```diff
@@ -9,16 +9,19 @@ module.exports = function(app, charted_root) {
...
+    var urlObj = url.parse(req.url, true)
 
-    if (!get_url) {
+    if (!urlObj.query.url) {
       res.statusCode = 400
       res.end('Bad Request: no url provided')
-      next()
```
Comment (valueof) [inline lib/charted.js]: Why did you remove `next()`?

File: lib/charted.js [hunk 3]
```diff
@@ -9,16 +9,19 @@ module.exports = function(app, charted_root) {
...
       return
     }
 
-    request(decodeURIComponent(get_url), function (err, resp, body) {
+    var get_url = urlObj.query.url.match(/:\/\//) ? urlObj.query.url :
```
Comment (valueof) [inline lib/charted.js]: Since `urlObj` is a result of `url.parse`, don't you have all the information in the object already? For example, instead of matching on the url, you can check whether `hostname` is `null`.

File: lib/charted.js [hunk 4]
```diff
@@ -9,16 +9,19 @@ module.exports = function(app, charted_root) {
...
+        ('http://' + req.headers.host + '/' +
+         urlObj.query.url.replace(/^\//, ''))
```
Comment (valueof) [inline lib/charted.js]: So here, instead of doing string concatenation and more regular expressions I think you can use `url.format` with a modified object:

```javascript
var urlObj = url.parse("/hello.txt")
urlObj.hostname = "medium.com"
urlObj.protocol = "http:"
url.format(urlObj);
> http://medium.com
```

Also, you should check whether the website is running on https or http.

**Comment by valueof** [comment]
First pass done.

**Comment by valueof** [comment]
We should start writing unit tests for charted.

File: lib/charted.js (hunk 1 above)
Comment (raykyri) [inline lib/charted.js]: I was trying to use urlObj for something else - I think your code in the other comment roughly reflects it.

File: lib/charted.js (hunk 2 above)
Comment (raykyri) [inline lib/charted.js]: Was getting warnings when other code that runs on the response object later tries to modify it. "Headers cannot be set after response is sent" or something similar.

The middleware is handling and returning for all requests that gets passed to it, so it doesn't seem like we should need a call to next().

File: lib/charted.js (hunk 3 above)
Comment (raykyri) [inline lib/charted.js]: See below...

File: lib/charted.js (hunk 4 above)
Comment (raykyri) [inline lib/charted.js]: In the second case, urlObj.query.url is just a relative path. It doesn't seem possible to get the protocol from urlObj and I couldn't find it elsewhere on the request or app objects when using Matador.

It isn't a problem in Express which is the server the Charted demo uses in this Github repo. But Connect (which Matador extends) doesn't seem to try to figure out how the user is talking to the web server, or those headers are stripped before they get passed to Charted.

Let me know if you think there's another way to figure out, or a header I haven't looked at.

**Comment by raykyri** [comment]
Thanks for the review.

File: lib/charted.js (hunk 4 above)
Comment (valueof) [inline lib/charted.js]: Well, if it's a relative path then url.Obj.hostname will be `null` (or empty), right? My example above handles relative part, no?

To make sure I'm not misunderstanding what you're trying to solve: sometimes you get relative URLs and you want to make them absolute by appending http://<hostname> in front? If this is the case, you can definitely use `url.format` and `url.parse` and you don't need regular expressions. We can chat about this next week, if you want.

File: lib/charted.js [hunk 5]
```diff
@@ -10,15 +10,16 @@ module.exports = function(app, charted_root) {
...
     if (!get_url) {
       res.statusCode = 400
       res.end('Bad Request: no url provided')
-      next()
       return
     }
 
-    request(decodeURIComponent(get_url), function (err, resp, body) {
+    var get_url_full = url.parse(get_url).host ? get_url :
+        ('http://' + req.headers.host + '/' + get_url.replace(/^\//, ''))
```
Comment (raykyri) [inline lib/charted.js]: @valueof: This seems to be as far as I can get in determining the absolute URL when this endpoint gets hit with a relative path, e.g. 'example.csv'. All of the ways I've looked at for obtaining the absolute URL seem to fail on Connect...

	console.log(req.protocol)
	console.log(req.url)
	console.log(req.originalUrl)
	console.log(req.host)
	console.log(req.headers.host)
	console.log(req.params)

You can verify this by changing index.js to initialize a Connect app instead of Express app.

Comment (chaosgame) [inline lib/charted.js]: Would https://nodejs.org/docs/latest/api/url.html#url_url_resolve_from_to help?

**Comment by valueof** [comment]
lgtm

**Comment by raykyri** [comment]
ayy, thanks
